src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -0,0 +1,333 @@
+import { ITelemetryService } from 'vs/platform/telemetry/common/telemetry';
+import { ILogService } from 'vs/platform/log/common/log';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { setTimeout0 } from 'vs/base/common/platform';
+import { importAMDNodeModule } from 'vs/amdX';
+
+const EDITOR_EXPERIMENTAL_PREFER_TREESITTER = 'editor.experimental.preferTreeSitter';
+const moduleLocationTreeSitter: AppResourcePath = `${nodeModulesPath}/@vscode/tree-sitter-wasm/wasm`;
+const moduleLocationTreeSitterWasm: AppResourcePath = `${moduleLocationTreeSitter}/tree-sitter.wasm`;
+
+class TextModelTreeSitter extends Disposable {
+	private _treeSitterTree: TreeSitterTree | undefined;
+	private _contentChangedListener: IDisposable | undefined;
+
+	constructor(readonly model: ITextModel,
+		private readonly _treeSitterParser: TreeSitterParser,
+		private readonly _treeSitterImporter: TreeSitterImporter,
+	) {
+		super();
+		this._register(this.model.onDidChangeLanguage(e => this._onDidChangeLanguage(e)));
+		this._onDidChangeLanguage({ newLanguage: this.model.getLanguageId() });
+	}
+
+	private _registerModelListeners() {
+		this._contentChangedListener = this.model.onDidChangeContent(e => this._onDidChangeContent(e));
+		this._onDidChangeContent();
+	}
+
+	private _unregisterModelListeners() {
+		this._contentChangedListener?.dispose();
+	}
+
+	private async _onDidChangeLanguage(e: { newLanguage: string }) {
+		if (this._treeSitterTree) {
+			this._treeSitterTree.dispose();
+		}
+		const language = await this._treeSitterParser.getLanguage(e.newLanguage);
+		if (!language || this.model.isDisposed()) {
+			this._unregisterModelListeners();
+			return;
+		}
+
+		const Parser = await this._treeSitterImporter.getParserClass();
+		this._treeSitterTree = new TreeSitterTree(new Parser(), language);
+		this._registerModelListeners();
+	}
+
+	private async _onDidChangeContent(e?: IModelContentChangedEvent) {
+		if (!this._treeSitterTree) {
+			return;
+		}
+
+		if (e) {
+			for (const change of e.changes) {
+				const newEndOffset = change.rangeOffset + change.text.length;
+				const newEndPosition = this.model.getPositionAt(newEndOffset);
+
+				this._treeSitterTree.tree?.edit({
+					startIndex: change.rangeOffset,
+					oldEndIndex: change.rangeOffset + change.rangeLength,
+					newEndIndex: change.rangeOffset + change.text.length,
+					startPosition: { row: change.range.startLineNumber - 1, column: change.range.startColumn - 1 },
+					oldEndPosition: { row: change.range.endLineNumber - 1, column: change.range.endColumn - 1 },
+					newEndPosition: { row: newEndPosition.lineNumber - 1, column: newEndPosition.column - 1 }
+				});
+			}
+		}
+
+		this._treeSitterTree.tree = await this._treeSitterParser.parse(this.model, this._treeSitterTree);
+	}
+
+	override dispose() {
+		super.dispose();
+		this._treeSitterTree?.dispose();
+		this._unregisterModelListeners();
+	}
+}
```

Comment:
I think an implementation like this would avoid most of the problems:

```suggestion
class TextModelTreeSitter extends Disposable {
	private _treeSitterTree: TreeSitterTree | undefined;

	constructor(readonly model: ITextModel,
		private readonly _treeSitterParser: TreeSitterParser,
		private readonly _treeSitterImporter: TreeSitterImporter,
	) {
		super();
		this._register(Event.runImmediatelyAndOnChange(this.model.onDidChangeLanguage, () => this._updateLanguage(this.model.getLanguageId())));
	}

	private readonly _sessionDisposables = this._register(new DisposableStore());

	private _updateLanguage(language: string): void {
		this._sessionDisposables.clear();
		this._treeSitterTree = undefined;

		const token = this._sessionDisposables.add(createCancelOnDisposeToken());
		const language = await this._treeSitterParser.getLanguage(e.newLanguage);
		if (token.isCancelled || !language) { return; }
		
		const parser = await this._treeSitterImporter.createParser();
		if (token.isCancelled) { return; }

		const tree = _sessionDisposables.add(new TreeSitterTree(parser, language));

		this._sessionDisposables.add(Event.runImmediatelyAndOnChange(this.model.onDidChangeContent, e => tree._parse(e?.changes)));
		
		this._treeSitterTree = tree;
	}
}
```